<a href="https://colab.research.google.com/github/AdityamKrovvidi/Coursera_Capstone/blob/main/AdityamW3Assign.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto

In [ ]:
! pip install BeautifulSoup4
from bs4 import BeautifulSoup
import requests

import pandas as pd
import numpy as np

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

print('Libraries imported.')

     |████████████████████████████████| 122kB 5.5MB/s eta 0:00:01
Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2020.12.5          |   py36h5fab9bb_1         143 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-2.1.0                |     pyhd3deb0d_0          64 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         240 KB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.50-py_0
  geopy              conda-forge/noarch::geopy-2.1.0-pyhd3deb0d_0

The following packages will be UPDATED:

  certifi                          2020.12.5-py36h5fab9bb_0

# 1. Create a Dataframe of Postal Codes of Canada

Scrape the Wikipedia page for postal codes in Canada

In [ ]:
#Get the page
URL = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
page = requests.get(URL)
soup = BeautifulSoup(page.content, 'html.parser')
#print(soup.prettify()) #commented to avoid lengthy output showing on GitHub without side scroll

In [ ]:
#Create an empty dataframe to fill the data later
canada_pc = pd.DataFrame(columns=['PinCode', 'Borough', 'Neighborhood'])
canada_pc

Empty DataFrame
Columns: [PinCode, Borough, Neighborhood]
Index: []

In [ ]:
#Using BeautifulSoup read the table data and copy it into the dataframe  
tb = soup.find('tbody') 
i= 0
for link in tb.find_all('tr'):
    j=0
    for name in link.find_all('td'):
        if (j == 0):
            name0 = name.get_text()    
        elif (j == 1):
            name1 = name.get_text()              
        else:
            name2 = name.get_text()         
        j=j+1
    
    if(i!=0):
        canada_pc = canada_pc.append({'PinCode' : name0, 'Borough' : name1, 'Neighborhood' : name2},  
               ignore_index = True)

    i=i+1

canada_pc = canada_pc.replace(r"\n","", regex=True) #To remove "\n" padded to every value in the dataframe
canada_pc.head()

PinCode           Borough               Neighborhood
0     M1A      Not assigned               Not assigned
1     M2A      Not assigned               Not assigned
2     M3A        North York                  Parkwoods
3     M4A        North York           Victoria Village
4     M5A  Downtown Toronto  Regent Park, Harbourfront

In [ ]:
canada_pc.shape

(180, 3)

Clean up the dataframe by removing rows where Borough is "Not assigned" and grouping duplicate postal codes

In [ ]:
#Remove rows where Borough is not assigned
i=0
canada = canada_pc
for boro in canada_pc['Borough']:
    if (boro == "Not assigned"):
        canada.drop(index=i, inplace=True)
    i=i+1
canada.reset_index(drop=True, inplace=True)
canada.head()

PinCode           Borough                                 Neighborhood
0     M3A        North York                                    Parkwoods
1     M4A        North York                             Victoria Village
2     M5A  Downtown Toronto                    Regent Park, Harbourfront
3     M6A        North York             Lawrence Manor, Lawrence Heights
4     M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [ ]:
canada.shape

(103, 3)

The rows are reduced from 180 to 103, i.e., 77 boroughs are removed

In [ ]:
#Check if there are any neighborhoods that are not assigned
canada[canada['Neighborhood'] =='Not assigned'].count()['Neighborhood']

0

In [ ]:
#Check if there are any duplicate postal codes
canada[canada.duplicated(subset='PinCode')=='True'].count()['PinCode']

0

# 2. Geocode the Postal Codes of Canada

Let's try Geocoder package

In [ ]:
# ! pip install geocoder
# import geocoder 

In [ ]:
# lat_lng_coords = None
# while(lat_lng_coords is None):
        
#     g = geocoder.google("M5G, Toronto, Ontario")
#     lat_lng_coords = g.latlng

# latitude = lat_lng_coords[0]
# longitude = lat_lng_coords[1]
# print("Lat= ", latitude, "  Long= ", longitude)

Geocoder failed to return latitude & longitude even for one postal code, so the following code also failed to work

In [ ]:
# for i in canada_pc.index:
#     #print(i)
#     postal_code = canada_pc.loc[i, 'PinCode']
#     boro = canada_pc.loc[i, 'Borough']
#     neigh = canada_pc.loc[i, 'Neighborhood']
# # initialize your variable to None
#     lat_lng_coords = None

# # loop until you get the coordinates
#     while(lat_lng_coords is None):
#         print(postal_code, boro, neigh)
#         g = geocoder.google('{}, {}, {}'.format(postal_code, boro, neigh))
#         lat_lng_coords = g.latlng

#     latitude = lat_lng_coords[0]
#     longitude = lat_lng_coords[1]

#     canada_pc = canada_pc.append({'Latitude': latitude, 'Longitude': longitude})
# canada_pc.head()

Let's try if Nominatim works

In [ ]:
# address = 'M5G, Downtown Toronto, Central Bay Street' 
# geolocator = Nominatim(user_agent="adityam_krovvidi@yahoo.co.in")
# try:
#     location = geolocator.geocode(address)
#     latitude = location.latitude
#     longitude = location.longitude
#     print('The geograpical coordinate are {}, {}.'.format(latitude, longitude))
# except:
#     print('Location not found')

Nominatim did not work for postal code in Canada but works for neighborhoods and boroughs. So, it does not help us. 

As a last resort I used the csv file provided.

In [ ]:
#Read the csv file containing the latitudes & longitudes
URL2 = 'https://cocl.us/Geospatial_data'
coordinates = pd.read_csv(URL2)
coordinates

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [ ]:
#Merge canada and coordinates to add latitude & longitude columns
coordinates.rename(columns={'Postal Code': 'PinCode'}, inplace=True)
canada_ll = canada
canada_ll = canada_ll.join(coordinates.set_index('PinCode'), on='PinCode')
canada_ll.head()

PinCode           Borough                                 Neighborhood  \
0     M3A        North York                                    Parkwoods   
1     M4A        North York                             Victoria Village   
2     M5A  Downtown Toronto                    Regent Park, Harbourfront   
3     M6A        North York             Lawrence Manor, Lawrence Heights   
4     M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Let us take a subset of canada with neighborhoods in Toronto for clustering 

In [ ]:
#Filter boroughs with strings containing Toronto and create a new dataframe
filt = canada_ll['Borough'].str.contains('Toronto', na=False)
toronto = canada_ll.loc[filt]
toronto.reset_index(drop=True, inplace=True)
toronto

PinCode           Borough  \
0      M5A  Downtown Toronto   
1      M7A  Downtown Toronto   
2      M5B  Downtown Toronto   
3      M5C  Downtown Toronto   
4      M4E      East Toronto   
5      M5E  Downtown Toronto   
6      M5G  Downtown Toronto   
7      M6G  Downtown Toronto   
8      M5H  Downtown Toronto   
9      M6H      West Toronto   
10     M5J  Downtown Toronto   
11     M6J      West Toronto   
12     M4K      East Toronto   
13     M5K  Downtown Toronto   
14     M6K      West Toronto   
15     M4L      East Toronto   
16     M5L  Downtown Toronto   
17     M4M      East Toronto   
18     M4N   Central Toronto   
19     M5N   Central Toronto   
20     M4P   Central Toronto   
21     M5P   Central Toronto   
22     M6P      West Toronto   
23     M4R   Central Toronto   
24     M5R   Central Toronto   
25     M6R      West Toronto   
26     M4S   Central Toronto   
27     M5S  Downtown Toronto   
28     M6S      West Toronto   
29     M4T   Central Toronto   
30     M5T  Downtown Toronto   
31     M4V   Central Toronto   
32     M5V  Downtown Toronto   
33     M4W  Downtown Toronto   
34     M5W  Downtown Toronto   
35     M4X  Downtown Toronto   
36     M5X  Downtown Toronto   
37     M4Y  Downtown Toronto   
38     M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                 North Toronto West,  Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
32  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
33                                           Rosedale  43.679563 -79.3

In [ ]:
print('The Toronto dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto['Borough'].unique()),
        toronto.shape[0]
    )
)

The Toronto dataframe has 4 boroughs and 39 neighborhoods.


Let's see the Toronto neighborhoods on a map

In [ ]:
#Get latitude & longitude of Toronto using Nominatim
address = 'Toronto, Ontario'
geolocator = Nominatim(user_agent="adityam_krovvidi@yahoo.co.in")
try:
    location = geolocator.geocode(address)
    latitude = location.latitude
    longitude = location.longitude
    print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))
except:
    print('Location not found')

The coordinates of Toronto are 43.6534817, -79.3839347.


In [ ]:
# create map of Toronto using latitude and longitude values of postal codes
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto['Latitude'], toronto['Longitude'], toronto['Borough'], toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto 
#If map is not visible on GitHub, please enter the GitHub URL in https://nbviewer.jupyter.org/ to view map  

# 3.0 Cluster Neighborhoods in Toronto

## 3.1 Explore venues around the first neighborhood in Toronto dataframe 

Let's use Foursqaure API to get venues around a neighborhood

In [ ]:
CLIENT_ID = 'QDXH54TPE0SA4WT0X4F3TK1KAH3XVGEGZI04LSC513SY3VYL' # Foursquare ID
CLIENT_SECRET = 'XR4YB4AYKR4SAOKGM0LLLSMTKB1FMMXO3OUJ4KVPWCCYQCYP' # Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: QDXH54TPE0SA4WT0X4F3TK1KAH3XVGEGZI04LSC513SY3VYL
CLIENT_SECRET:XR4YB4AYKR4SAOKGM0LLLSMTKB1FMMXO3OUJ4KVPWCCYQCYP


In [ ]:
#Create a URL for the first neighborhood in Toronto dataframe
latitude = toronto.loc[0, 'Latitude']
longitude = toronto.loc[0, 'Longitude']
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=QDXH54TPE0SA4WT0X4F3TK1KAH3XVGEGZI04LSC513SY3VYL&client_secret=XR4YB4AYKR4SAOKGM0LLLSMTKB1FMMXO3OUJ4KVPWCCYQCYP&ll=43.6542599,-79.3606359&v=20180605&radius=500&limit=100'

In [ ]:
import json # library to handle JSON files
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [ ]:
#Get the json file of venues 
results = requests.get(url).json()
#results #commented to avoid lengthy output showing on GitHub without side scroll

In [ ]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories'] # useful when not sure where 'categories' is in the hierarchy 
    except:
        categories_list = row['venue.categories'] # 'categories' is actually in 'venue'
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [ ]:
from IPython.display import display # to display wider dataframes neatly
venues = results['response']['groups'][0]['items']
#venues #commented to avoid lengthy output showing on GitHub without side scroll

In [ ]:
nearby_venues = json_normalize(venues) # flatten JSON, i.e. convert list to dataframe with keys as columns
display(nearby_venues.head())

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  """Entry point for launching an IPython kernel.


referralId  reasons.count  \
0  e-0-53b8466a498e83df908c3f21-0              0   
1  e-0-54ea41ad498e9a11e9e13308-1              0   
2  e-0-574c229e498ebb5c6b257902-2              0   
3  e-0-50760559e4b0e8c7babe2497-3              0   
4  e-0-5612b1cc498e3dd742af0dc8-4              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   
4  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id              venue.name venue.location.address  \
0  53b8466a498e83df908c3f21           Tandem Coffee          368 King St E   
1  54ea41ad498e9a11e9e13308        Roselle Desserts          362 King St E   
2  574c229e498ebb5c6b257902  Cooper Koo Family YMCA          461 Cherry St   
3  50760559e4b0e8c7babe2497     Body Blitz Spa East   497 King Street East   
4  5612b1cc498e3dd742af0dc8          Impact Kitchen          573 King St E   

        venue.location.crossStreet  venue.location.lat  venue.location.lng  \
0                    at Trinity St           43.653559          -79.361809   
1                       Trinity St           43.653447          -79.362017   
2                              NaN           43.653249          -79.358008   
3  btwn Sackville St and Sumach St           43.654735          -79.359874   
4                at St Lawrence St           43.656369          -79.356980   

                       venue.location.labeledLatLngs  ...  \
0  [{'label': 'display', 'lat': 43.65355870959944...  ...   
1  [{'label': 'display', 'lat': 43.65344672305267...  ...   
2  [{'label': 'display', 'lat': 43.65324910177244...  ...   
3  [{'label': 'display', 'lat': 43.65473505045365...  ...   
4  [{'label': 'display', 'lat': 43.65636850543279...  ...   

   venue.location.city venue.location.state venue.location.country  \
0              Toronto                   ON                 Canada   
1              Toronto                   ON                 Canada   
2              Toronto                   ON                 Canada   
3              Toronto                   ON                 Canada   
4              Toronto                   ON                 Canada   

                     venue.location.formattedAddress  \
0  [368 King St E (at Trinity St), Toronto ON, Ca...   
1  [362 King St E (Trinity St), Toronto ON M5A 1K...   
2        [461 Cherry St, Toronto ON M5A 0H7, Canada]   
3  [497 King Street East (btwn Sackville St and S...   
4  [573 King St E (at St Lawrence St), Toronto ON...   

                                    venue.categories venue.photos.count  \
0  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...                  0   
1  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...                  0   
2  [{'id': '52e81612bcbc57f1066b7a37', 'name': 'D...                  0   
3  [{'id': '4bf58dd8d48988d1ed941735', 'name': 'S...                  0   
4  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...                  0   

  venue.photos.groups  venue.location.postalCode venue.venuePage.id  \
0                  []                        NaN                NaN   
1                  []                    M5A 1K9                NaN   
2                  []                    M5A 0H7                NaN   
3                  []                    M5A 1L9                NaN   
4                  []                    M5A 4L3                NaN   

  venue.location.neighborhood  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN  

[5 rows x 22 columns]

In [ ]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

display(nearby_venues.head())

name           categories        lat        lng
0           Tandem Coffee          Coffee Shop  43.653559 -79.361809
1        Roselle Desserts               Bakery  43.653447 -79.362017
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3     Body Blitz Spa East                  Spa  43.654735 -79.359874
4          Impact Kitchen           Restaurant  43.656369 -79.356980

In [ ]:
nearby_venues.shape

(43, 4)

## 3.2 Explore all neighborhoods in Toronto

Let's define a function to repeat the same process to all the neighborhoods

In [ ]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
    
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list]) # list is converted to a dataframe
    # venue_list is first assigned an element (a neighborhood) from the venues_list and then item is assigned elements from venue_list (touples of venues)
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    # nearby_venues is a dataframe of venues_list
    nearby_venues
    return(nearby_venues)

Let's run the above function on each neighborhood and create a new dataframe called toronto_venues

In [ ]:
pd.set_option('display.width', 200)
toronto_venues = getNearbyVenues(names=toronto['Neighborhood'],
                                   latitudes=toronto['Latitude'],
                                   longitudes=toronto['Longitude'])
toronto_venues

Neighborhood  Neighborhood Latitude  Neighborhood Longitude                    Venue  Venue Latitude  Venue Longitude        Venue Category
0                             Regent Park, Harbourfront              43.654260              -79.360636            Tandem Coffee       43.653559       -79.361809           Coffee Shop
1                             Regent Park, Harbourfront              43.654260              -79.360636         Roselle Desserts       43.653447       -79.362017                Bakery
2                             Regent Park, Harbourfront              43.654260              -79.360636   Cooper Koo Family YMCA       43.653249       -79.358008   Distribution Center
3                             Regent Park, Harbourfront              43.654260              -79.360636      Body Blitz Spa East       43.654735       -79.359874                   Spa
4                             Regent Park, Harbourfront              43.654260              -79.360636           Impact Kitchen       43.656369       -79.356980            Restaurant
...                                                 ...                    ...                     ...                      ...             ...              ...                   ...
1596  Business reply mail Processing Centre, South C...              43.662744              -79.321558     TTC Russell Division       43.664908       -79.322560    Light Rail Station
1597  Business reply mail Processing Centre, South C...              43.662744              -79.321558  Jonathan Ashbridge Park       43.664702       -79.319898                  Park
1598  Business reply mail Processing Centre, South C...              43.662744              -79.321558       Toronto Yoga Mamas       43.664824       -79.324335           Yoga Studio
1599  Business reply mail Processing Centre, South C...              43.662744              -79.321558              ONE Academy       43.662253       -79.326911  Gym / Fitness Center
1600  Business reply mail Processing Centre, South C...              43.662744              -79.321558     Revolution Recording       43.662561       -79.326940      Recording Studio

[1601 rows x 7 columns]

In [ ]:
#Let's check how many venues were returned for each neighborhood
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  Venue Latitude  Venue Longitude  Venue Category
Neighborhood                                                                                                                                             
Berczy Park                                                            56                      56     56              56               56              56
Brockton, Parkdale Village, Exhibition Place                           23                      23     23              23               23              23
Business reply mail Processing Centre, South Ce...                     16                      16     16              16               16              16
CN Tower, King and Spadina, Railway Lands, Harb...                     17                      17     17              17               17              17
Central Bay Street                                                     63                      63     63              63               63              63
Christie                                                               16                      16     16              16               16              16
Church and Wellesley                                                   77                      77     77              77               77              77
Commerce Court, Victoria Hotel                                        100                     100    100             100              100             100
Davisville                                                             34                      34     34              34               34              34
Davisville North                                                        8                       8      8               8                8               8
Dufferin, Dovercourt Village                                           14                      14     14              14               14              14
First Canadian Place, Underground city                                100                     100    100             100              100             100
Forest Hill North & West, Forest Hill Road Park                         4                       4      4               4                4               4
Garden District, Ryerson                                              100                     100    100             100              100             100
Harbourfront East, Union Station, Toronto Islands                     100                     100    100             100              100             100
High Park, The Junction South                                          24                      24     24              24               24              24
India Bazaar, The Beaches West                                         21                      21     21              21               21              21
Kensington Market, Chinatown, Grange Park                              63                      63     63              63               63              63
Lawrence Park                                                           3                       3      3               3                3               3
Little Portugal, Trinity                                               42                      42     42              42               42              42
Moore Park, Summerhill East                                             2                       2      2               2                2               2
North Toronto West,  Lawrence Park                                     17                      17     17              17               17              17
Parkdale, Roncesvalles                                                 14                      14     14              14               14              14
Queen's Park, Ontario Provincial Government                            33                      33     33              33               33              33
Regent Park, Harbourfront                       

In [ ]:
#Let's find out how many unique categories can be curated from all the returned venues
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 236 unique categories.


## 3.3 Analyze each neighborhood

In [ ]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  Theater  Theme Restaurant  \
0            0        0                   0             0               0                0                 0                    0             0         0  ...        0                 0   
1            0        0                   0             0               0                0                 0                    0             0         0  ...        0                 0   
2            0        0                   0             0               0                0                 0                    0             0         0  ...        0                 0   
3            0        0                   0             0               0                0                 0                    0             0         0  ...        0                 0   
4            0        0                   0             0               0                0                 0                    0             0         0  ...        0                 0   

   Tibetan Restaurant  Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  Wine Bar  
0                   0                 0      0              0                              0                 0                      0         0  
1                   0                 0      0              0                              0                 0                      0         0  
2                   0                 0      0              0                              0                 0                      0         0  
3                   0                 0      0              0                              0                 0                      0         0  
4                   0                 0      0              0                              0                 0                      0         0  

[5 rows x 236 columns]

In [ ]:
toronto_onehot.shape

(1601, 236)

In [ ]:
#let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                                         Berczy Park     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
2   Business reply mail Processing Centre, South C...     0.062500  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.058824            0.058824      0.058824        0.117647         0.176471          0.117647             0.000000      0.000000   
4                                  Central Bay Street     0.015873  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
5                                            Christie     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
6                                Church and Wellesley     0.025974  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.012987      0.000000   
7                      Commerce Court, Victoria Hotel     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.040000      0.000000   
8                                          Davisville     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
9                                    Davisville North     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
10                       Dufferin, Dovercourt Village     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
11             First Canadian Place, Underground city     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.030000      0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
13                           Garden District, Ryerson     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
15                      High Park, The Junction South     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.041667   
16                     India Bazaar, The Beaches West     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
18                                      Lawrence Park     0.000000  0.000000            0.000000      0.000000        0.000000         0.000000          0.000000             0.000000      0.000000   
19                           Little Portu

In [ ]:
#Write a function to sort the venues in descending order.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [ ]:
#Now let's create the new dataframe and display the top 10 venues for each neighborhood.
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                                        Berczy Park           Coffee Shop          Cocktail Bar                Bakery              Beer Bar        Farmers Market    Seafood Restaurant   
1       Brockton, Parkdale Village, Exhibition Place                  Café           Coffee Shop        Breakfast Spot             Nightclub                   Gym               Stadium   
2  Business reply mail Processing Centre, South C...           Yoga Studio         Garden Center            Restaurant         Burrito Place         Auto Workshop  Fast Food Restaurant   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service        Airport Lounge      Airport Terminal           Coffee Shop       Harbor / Marina   Rental Car Location   
4                                 Central Bay Street           Coffee Shop        Sandwich Place    Italian Restaurant                  Café       Bubble Tea Shop          Burger Joint   

  7th Most Common Venue 8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0           Cheese Shop            Restaurant                 Bistro              Jazz Club  
1         Burrito Place            Restaurant           Climbing Gym              Pet Store  
2        Farmers Market               Brewery       Recording Studio                 Garden  
3      Sculpture Garden                 Plane          Boat or Ferry               Boutique  
4           Salad Place            Poke Place  Portuguese Restaurant       Ramen Restaurant

## 3.4 Cluster neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [ ]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:40] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 1, 0,
       0, 0, 0, 0, 2, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [ ]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() 

PinCode           Borough                                 Neighborhood   Latitude  Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0     M5A  Downtown Toronto                    Regent Park, Harbourfront  43.654260 -79.360636               0           Coffee Shop                  Park                Bakery   
1     M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301 -79.389494               0           Coffee Shop      Sushi Restaurant                   Gym   
2     M5B  Downtown Toronto                     Garden District, Ryerson  43.657162 -79.378937               0        Clothing Store           Coffee Shop    Italian Restaurant   
3     M5C  Downtown Toronto                               St. James Town  43.651494 -79.375418               0                  Café           Coffee Shop          Cocktail Bar   
4     M4E      East Toronto                                  The Beaches  43.676357 -79.293031               0           Coffee Shop      Asian Restaurant                 Trail   

  4th Most Common Venue 5th Most Common Venue      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Breakfast Spot                   Pub                       Café               Theater            Shoe Store               Brewery             Restaurant  
1              Creperie                  Park                Music Venue    Mexican Restaurant   Japanese Restaurant    Italian Restaurant    Fried Chicken Joint  
2   Japanese Restaurant       Bubble Tea Shop  Middle Eastern Restaurant        Cosmetics Shop                  Café        Lingerie Store   Fast Food Restaurant  
3   American Restaurant             Gastropub                        Gym   Moroccan Restaurant        Lingerie Store    Seafood Restaurant     Italian Restaurant  
4                   Pub     Health Food Store                   Wine Bar               Dog Run                 Diner        Discount Store    Distribution Center

Finally, let's visualize the resulting clusters

In [ ]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters
#If map is not visible on GitHub, please enter the GitHub URL in https://nbviewer.jupyter.org/ to view map  

## 3.4 Examine clusters

Let's examine each cluster and determine the discriminating venue categories that distinguish each cluster. Since clusters 1, 2, 3 & 4 have only one neighborhood each there must be some venues distinguishing them from the cluster 0. Let's look at them.

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
29  Central Toronto               1                Lawyer                 Trail      Department Store           Event Space  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29   Dumpling Restaurant            Donut Shop       Doner Restaurant

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
33  Downtown Toronto               2                  Park                 Trail            Playground         Deli / Bodega  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
33   Dumpling Restaurant            Donut Shop       Doner Restaurant

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
19  Central Toronto               3  Fast Food Restaurant          Home Service                Garden         Deli / Bodega  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19   Dumpling Restaurant            Donut Shop       Doner Restaurant

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
18  Central Toronto               4                  Park              Bus Line           Swim School           Event Space     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18            Donut Shop      Doner Restaurant                Dog Run

Cluster 1 is unique with Lawyer being the most visited. Cluster 2 is for open spaces with Park, trail and playground being the most visited. Cluster 3 is unique for Home Service. Cluster 4 is unique with Bus Line and Swim School. So, I would classify the clusters having the following features that distinguish them from others:

* Cluster 0 - General (Coffee, Pubs, Restaurants, Shops, etc.)
* Cluster 1 - Legal, Department Store and Events
* Cluster 2 - Open and Green Spaces
* Cluster 3 - Home Service and Garden
* Cluster 4 - Bus Line, Swim School and Park

## *** END OF ASSIGNMENT ***